### PR TITLE
Fix index locator cast bool key to float.

### LINKED ISSUE #19087: index locators cast bool to float (and not to int)
#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.Index([0, 1, 2, 3]).get_loc(True) # right
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/home/pietro/nobackup/repo/pandas/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2513             try:
-> 2514                 return self._engine.get_loc(key)
   2515             except KeyError:

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5834)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5639)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index_class_helper.pxi in pandas._libs.index.Int64Engine._check_type (pandas/_libs/index.c:18970)()

KeyError: True

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-2-31d89f73ae5f> in <module>()
----> 1 pd.Index([0, 1, 2, 3]).get_loc(True)

/home/pietro/nobackup/repo/pandas/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2514                 return self._engine.get_loc(key)
   2515             except KeyError:
-> 2516                 return self._engine.get_loc(self._maybe_cast_indexer(key))
   2517 
   2518         indexer = self.get_indexer([key], method=method, tolerance=tolerance)

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5834)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5639)()

/home/pietro/nobackup/repo/pandas/pandas/_libs/index_class_helper.pxi in pandas._libs.index.Int64Engine._check_type (pandas/_libs/index.c:18970)()

KeyError: True

In [3]: pd.Index([0, 1, 2, 3.]).get_loc(True) # wrong
Out[3]: 1
```

#### Problem description

Given that bools are not casted to ints (which I think is good), even less one would expect that they are casted to float.

#### Expected Output

``KeyError: True`` in both cases.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-4-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.22.0.dev0+448.gdbec3c92e
pytest: 3.0.6
pip: 9.0.1
setuptools: 33.1.1
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.18.1
pyarrow: None
xarray: None
IPython: 5.2.2
sphinx: None
patsy: 0.4.1+dev
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: None
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by machenity**
- closes #19087

### CODE REVIEW

**Comment by gfyoung** [comment]
@machenity : Good start!  We're going to need tests and a `whatsnew` entry in `0.24.0`.

**Comment by machenity** [comment]
@gfyoung Do your 'tests' mean that i have to write new test case code or just be okay with those codecov test?

**Comment by scari** [comment]
I believe @gfyoung meant a test for your change. Will you come to the PyCon KR sprint tomorrow?

**Comment by machenity** [comment]
@scari Yes, i will, but i have a tutorial before that.

File: pandas/core/indexes/numeric.py [hunk 1]
```diff
@@ -403,7 +403,7 @@ def __contains__(self, other):
     @Appender(_index_shared_docs['get_loc'])
     def get_loc(self, key, method=None, tolerance=None):
         try:
-            if np.all(np.isnan(key)):
+            if np.all(np.isnan(key)) or isinstance(key, bool):
```
Comment (jreback) [inline pandas/core/indexes/numeric.py]: use is_bool here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/numeric.py (hunk 1 above)
Comment (machenity) [reply pandas/core/indexes/numeric.py]: Does python have `is_bool()` function?
I used `isinstance()` to check if key is an instance of subclass of bool.

Comment (machenity) [reply pandas/core/indexes/numeric.py]: Oh sorry, I found it.

Comment (jreback) [reply pandas/core/indexes/numeric.py]: pandas has all of these functions, which we consistently use. thanks

**Comment by machenity** [comment]
@gfyoung I added tests and `whatsnew` changes and edit codes with help of @jreback. Thanks.

File: doc/source/whatsnew/v0.24.0.txt [hunk 2]
```diff
@@ -653,6 +653,7 @@ Indexing
 - Fixed ``DataFrame[np.nan]`` when columns are non-unique (:issue:`21428`)
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
 - Bug where indexing with a Numpy array containing negative values would mutate the indexer (:issue:`21867`)
+- ``Float64Index.get_loc`` now raises KeyError when boolean key passed. (:issue:`19087`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.txt]: Double back-ticks on KeyError.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. small change suggested by @gfyoung

**Comment by machenity** [comment]
@gfyoung I edited it. Thank you for your suggestion.

**Comment by jreback** [comment]
thanks @machenity

**Pull request merged**